lib/compiler/src/beam_doc.erl

```diff
@@ -0,0 +1,285 @@
+            %% First we check if there is a doc prototype
+            case extract_slogan(Doc, F, A) of
+                undefined -> {io_lib:format("~p/~p",[F,A]), Doc};
+                SloganDocs -> SloganDocs
+            end,
+        AttrBody = {function, F, A},
+        FunDoc = gen_doc(Anno, AttrBody, Slogan, DocsWithoutSlogan, State),
+        [FunDoc | extract_documentation(T, reset_state(State))]
+    else
+        _ -> extract_documentation(T, reset_state(State))
+    end;
+extract_documentation([{attribute, Anno, TypeOrOpaque, {Type, _, TypeArgs}}=_AST | T],
+                      #docs{exported_types=ExpTypes}=State)
+  when TypeOrOpaque =:= type orelse TypeOrOpaque =:= opaque ->
+    maybe
+        Args = fun_to_varargs(TypeArgs),
+        true ?= sets:is_element({Type, length(Args)}, ExpTypes),
+        FunDoc = template_gen_doc({type, Anno, Type, length(Args), Args}, State),
+        [FunDoc | extract_documentation(T, reset_state(State))]
+    else
+        _ -> extract_documentation(T, reset_state(State))
+    end;
+extract_documentation([{attribute, Anno, callback, {{CB, A}, [Fun]}}=_AST | T], State) ->
+    Args = fun_to_varargs(Fun),
+    FunDoc = template_gen_doc({callback, Anno, CB, A, Args}, State),
+    [FunDoc | extract_documentation(T, reset_state(State))];
+extract_documentation([_H|T], State) ->
+    extract_documentation(T, State);
+extract_documentation([], #docs{doc = none}) ->
+    [].
+
+
+-spec gen_doc(Anno, AttrBody, Slogan, Docs, State) -> Response when
+      Anno      :: erl_anno:anno(),
+      AttrBody  :: {function | type | callback, term(), integer()},
+      Slogan    :: unicode:chardata(),
+      Docs      :: none | hidden | unicode:chardata(),
+      State     :: internal_docs(),
+      Signature :: [binary()],
+      D         :: map() | none | hidden,
+      Meta      :: map(),
+      Response  :: {AttrBody, Anno, Signature, D, Meta}.
+gen_doc(Anno0, AttrBody, Slogan, Doc, #docs{meta = Meta, module_name = Module})
+    when Doc =:= none orelse Doc =:= hidden ->
+    Anno1 = erl_anno:set_file(Module, Anno0),
+    {AttrBody, Anno1, [unicode:characters_to_binary(Slogan)], Doc, Meta};
+gen_doc(Anno0, AttrBody, Slogan, Docs, #docs{meta = Meta, module_name = Module}) ->
+    Anno1 = erl_anno:set_file(Module, Anno0),
+    {AttrBody, Anno1, [unicode:characters_to_binary(Slogan)],
+      #{ <<"en">> => unicode:characters_to_binary(string:trim(Docs)) }, Meta}.
+
+template_gen_doc({Attr, Anno, F, A, _Args}, #docs{doc = Doc}=State) when Doc =:= none ->
+    {Slogan, DocsWithoutSlogan} = {io_lib:format("~p/~p",[F,A]), Doc},
+    AttrBody = {Attr, F, A},
+    gen_doc(Anno, AttrBody, Slogan, DocsWithoutSlogan, State);
+template_gen_doc({Attr, Anno, F, A, Args}, #docs{doc = Doc}=State) ->
+    {Slogan, DocsWithoutSlogan} =
+        case extract_slogan(Doc, F, A) of
+            undefined ->
+                maybe
+                    true ?= lists:all(fun({var,_,N}) when N =/= '_' -> true; (_) -> false end, Args),
```

Comment:
The convention in the compiler application is to import functions from the `lists` module.

Reply:
Yes, I forgot that.